src/unix/process.c

```diff
@@ -342,6 +346,274 @@ static void uv__process_child_init(const uv_process_options_t* options,
 }
 #endif
 
+#if defined(__APPLE__)
+
+typedef struct uv__posix_spawn_fncs_tag {
+  struct {
+    int (*set_uid_np)(const posix_spawnattr_t *, uid_t);
+    int (*set_gid_np)(const posix_spawnattr_t *, gid_t);
+    int (*set_groups_np)(const posix_spawnattr_t*, int, gid_t*, uid_t);
+  } spawnattr;
+
+  struct {
+    int (*addchdir_np)(const posix_spawn_file_actions_t *, const char *);
+  } file_actions;
+} uv__posix_spawn_fncs_t;
+
+int uv__spawn_use_posix_spawn(uv__posix_spawn_fncs_t* fncs) {
+  /* Try to locate all non-portable functions at runtime */
+  fncs->spawnattr.set_uid_np = dlsym(RTLD_DEFAULT, "posix_spawnattr_set_uid_np");
+  fncs->spawnattr.set_gid_np = dlsym(RTLD_DEFAULT, "posix_spawnattr_set_gid_np");
+  fncs->spawnattr.set_groups_np = dlsym(RTLD_DEFAULT, "posix_spawnattr_set_groups_np");
+  fncs->file_actions.addchdir_np = dlsym(RTLD_DEFAULT, "posix_spawn_file_actions_addchdir_np");
+
+  /* If any of these is missing, just don't use posix_spawn */
+  return fncs->spawnattr.set_gid_np != NULL && 
+         fncs->spawnattr.set_uid_np != NULL && 
+         fncs->spawnattr.set_groups_np != NULL &&
+         fncs->file_actions.addchdir_np != NULL;
+}
+
+int uv__spawn_set_posix_spawn_attrs(posix_spawnattr_t* attrs,
+                                    const uv__posix_spawn_fncs_t* posix_spawn_fncs,
+                                    const uv_process_options_t* options)
+{
+  int err;
+  unsigned int flags;
+  sigset_t signal_set;
+
+  err = posix_spawnattr_init(attrs);
+  if (err != 0) {
+    /* If initialization fails, no need to de-init, just return */
+    return err;
+  }
+
+  if (options->flags & UV_PROCESS_SETUID) {
+    err = posix_spawn_fncs->spawnattr.set_uid_np(attrs, options->uid);
+    if (err != 0)
+      goto error;
+  }
+
+  if (options->flags & UV_PROCESS_SETGID) {
+    err = posix_spawn_fncs->spawnattr.set_gid_np(attrs, options->gid);
+    if (err != 0) 
+      goto error;
+  }
+
+  if (options->flags & (UV_PROCESS_SETUID | UV_PROCESS_SETGID)) {
+    /* See the comment on the call to setgroups in uv__process_child_init above
+      * for why this is not a fatal error */
+    SAVE_ERRNO(posix_spawn_fncs->spawnattr.set_groups_np(attrs, 0, NULL, KAUTH_UID_NONE));
```

Comment:
Is there any documentation on this? The best I could find is the source code, which indicates that NULL is invalid:

https://github.com/Hardercoder/opensourceapple-tarballs/blob/235bdeaf60d803fb88de0da8d5a53d67e7eaafd0/xnu-6153.11.26/libsyscall/wrappers/spawn/posix_spawn.c#L2193

Reply:
None that I could find, hence the caveat in the description. I used the source for reference too, but failed to check that restriction. The ngroups set to zero means that that var will never be read (besides the if call, of course), so I think we can just pass in a pointer to a gid_t and that should be enough...